openedx/core/djangoapps/content_libraries/docs/decisions/0004-upstream-block.rst

```diff
@@ -0,0 +1,89 @@
+
+Pending implementation in (TODO - link)
+
+Context
+*******
+
+* TODO - context
+
+Decision
+********
+
+* The following two fields will be added to every XBlock in CMS:
+
+  .. code-block:: python
+
+     upstream_block = String(
+         scope=Scope.settings,
+         help=(
+             "The usage key of a the block that served as this block's template, if one exists. "
+             "Generally, the upstream_block is within a Content Library."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+     upstream_block_version = Integer(
+         scope=Scope.settings,
+         help=(
+             "The upstream_block's version number, at the time this block was created from it. "
+             "If this version is older than the upstream_block's latest version, then CMS will "
+             "allow this block to fetch updated content from upstream_block."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+
+* The upstream-defined values of XBlock settings will be exposed in the XBlock API (TODO - details)
+
+* The upstream-defined values will be set when the XBlock is initialized by (TODO - figure this out)
+
+* The existing ``get_block_original_usage`` Modulestore methods will (TODO - figure this out)
+
+* New attributes will be serialized into the OLX of blocks with upstreams:
+
+  * The two new XBlock fields will serialize into OLX as attributes, like any settings-scoped field.
+  * The upstream defaults will use a new special syntax: ``upstream.$FIELD_NAME``.
+  * The period (``.``) was explicitly chosen to ensure that these special attributes cannot possibly conflict with any Python-defined XBlock fields.
+
+    .. code-block:: xml
+
+       <problem
+         display_name="A title that has been customized in the course"
+         max_attempts=2
+         upstream_block="lb:myorg:mylib:problem:p1"
+         upstream_block_version="12"
+         upstream_block.display_name="The title that was defined in the library block"
+         upstream_block.max_attempts=3
+       >
+       <!-- problem content would go here -->
+       </problem>
+
+* Existing LibraryContentBlock children will be missing these attributes, and we will need to handle that (TODO - more details)
+
+Consequences
+************
+
+* LibraryContentBlock consequences:
+
+  * Previous ADR (TODO - link) on LibraryContentBlock schema is null and void.
+  * Statically-referenced library content will be direct children of the Unit with no LibraryContentBlock wrapper.
+  * LibraryContentBlock will only be used for V1 libraries and V2 randomized problem banks.
+  * Eventually, we will deprecate V1 libraries and/or port them to V2 randomized problem banks.
+  * Long-term, we will remove LibraryContentBlock in favor of a Unit compositor.
+  * However, the ``<library_content>`` OLX tag will still be used for randomized content. Would be nice to rename this to ``<randomized>`` ?
+
+* Course-library interaction consequences:
```

Comment:
I think this approach also leaves copy-paste in an odd place when it comes to copying something from a Library and Pasting it into a course–i.e. should the pasting be smart enough to make the connection between the two and transform all the fields into `upstream.{}` fields? Not blocking, and it depends on product input, but it might be worth mentioning here.

Reply:
We could prompt the user when they paste from a library. "Do you want to paste this as linked content, that stays in sync with the original library version?"

Reply:
Good point. I'll note that possibility here.

Reply:
Related: for "regular" (non-randomized) blocks linked from a library into a course that have the `upstream_block` attribute, it would now be very easy to provide a method for users to "unlink" them from the library and convert them to normal, editable course blocks. We simply remove the `upstream_block` attribute. (For content which is a child of a LibraryContentBlock/ItemPool/ProblemBank, it's not as simple though.)

Reply:
👍🏻 I'll mention all these as potential ways that the product could use the new fields